sql/server.go

```diff
@@ -153,37 +153,45 @@ func (p parameters) Arg(i int) (parser.Datum, bool) {
 
 // exec executes the request. Any error encountered is returned; it is
 // the caller's responsibility to update the response.
-func (s *Server) exec(req driver.Request) (driver.Response, error) {
-	var resp driver.Response
+func (s *Server) exec(req driver.Request) (resp driver.Response, err error) {
 
 	// Pick up current session state.
 	// The request user is validated in ServeHTTP. Even in insecure mode,
 	// it is guaranteed not to be empty.
-	planner := planner{user: req.GetUser()}
+	planMaker := planner{user: req.GetUser()}
+	defer func(p *planner) {
```

Comment:
We currently already have two methods ServeHTTP and exec(), where an error found in ServeHTTP doesn't result in session state being set by the server. But I do agree the use of defer makes one have to look at all the error scenarios and wonder if they all should legitimately allow session to be sent back.